File: src/demo/MInStack_2stacks.java
package demo;

import java.util.Stack;



//Time complexity : O(1)
//Space Complexity: O(n)
//LeetCode : yes

public class MInStack_2stacks {

}


class MinStack {
//ONE TO ONE RELATIONSHIP BETWEEN 2 STACKS
    Stack<Integer> st = new Stack<>() ;
    Stack <Integer> minStack = new Stack<>();
    Integer min = Integer.MAX_VALUE;
    /** initialize your data structure here. */
    public MinStack() {
        minStack.push(min);
    }
    
    public void push(int x) {
        st.push(x);
        if(x <= min ){
            min = x;
        }
        minStack.push(min);
    }
    
    public void pop() {
        st.pop();
        minStack.pop();
        min = minStack.peek();
    }
    
    public int top() {
        return st.peek();
    }
    
    public int getMin() {
        return min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: MinStack_2Stacks.py
class MinStack:
    

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.sMain = [] #Main Stack
        self.sMin = [] #Min Stack
        self.minimum = sys.maxsize#2147483647 #Max Value of python
        self.sMin.append(self.minimum)
    

    def push(self, x: int) -> None:
        self.sMain.append(x)
        self.minimum = min(self.minimum,x)
        self.sMin.append(self.minimum)
        

    def pop(self) -> None:
        self.sMain.pop() # pop without index will give the last value in the index
        self.sMin.pop() # pop from both stacks: refer notes
        self.minimum = self.sMin[-1] #peek() function, now the min value will                                                   be the previous one

    def top(self) -> int:
        return self.sMain[-1]#return the last value in the list, which is the top                                value                                 when looked in a stack perspective; peek()
        

    def getMin(self) -> int:
        return self.minimum
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()